File: Julia/bert-segment-embedding.jl
mutable struct BertEmbeddings
    hidden_size::Int
    token_embeddings
    position_embeddings
    segment_embeddings
end

function BertEmbeddings(vocab_size::Int, max_position::Int, hidden_size::Int)
    token_embeddings = randn(vocab_size, hidden_size) .* 0.02
    position_embeddings = randn(max_position, hidden_size) .* 0.02
    segment_embeddings = randn(2, hidden_size) .* 0.02
    BertEmbeddings(hidden_size, token_embeddings, position_embeddings, segment_embeddings)
end

function forward(emb::BertEmbeddings, token_ids, segment_ids)
    tok_emb = emb.token_embeddings[token_ids .+ 1, :]
    seq_len = size(token_ids, 2)
    positions = 1:seq_len
    pos_emb = emb.position_embeddings[positions, :]
    seg_emb = emb.segment_embeddings[segment_ids .+ 1, :]
    tok_emb .+ pos_emb .+ seg_emb
end

File: R/bert-segment-embedding.R
BertEmbeddings <- setRefClass(
  "BertEmbeddings",
  fields = list(
    hidden_size = "numeric",
    token_embeddings = "matrix",
    position_embeddings = "matrix",
    segment_embeddings = "matrix"
  ),
  methods = list(
    initialize = function(vocab_size, max_position, hidden_size) {
      hidden_size <<- hidden_size
      token_embeddings <<- matrix(rnorm(vocab_size * hidden_size, sd = 0.02), nrow = vocab_size, ncol = hidden_size)
      position_embeddings <<- matrix(rnorm(max_position * hidden_size, sd = 0.02), nrow = max_position, ncol = hidden_size)
      segment_embeddings <<- matrix(rnorm(2 * hidden_size, sd = 0.02), nrow = 2, ncol = hidden_size)
    },
    forward = function(token_ids, segment_ids) {
      tok_emb <- token_embeddings[token_ids + 1, , drop = FALSE]
      seq_len <- ncol(token_ids)
      positions <- 1:seq_len
      pos_emb <- position_embeddings[positions, , drop = FALSE]
      seg_emb <- segment_embeddings[segment_ids + 1, , drop = FALSE]
      tok_emb + pos_emb + seg_emb
    }
  )
)
